### PR TITLE
💅 Add dynamic highlighting when scrolling Table of Contents

### PR DESCRIPTION
**Pull request by MeganSteenkamp**
Display table of contents text in a brand accent colour when the section is visible within the viewport.

**Test plan**: Expect to scroll and see dynamic highlighting of table of contents.

Screenshot 2023-01-11 at 1 16 18 PM
Screenshot 2023-01-11 at 1 16 06 PM

### CODE REVIEW

File: components/Layout.vue [hunk 1]
```diff
@@ -93,6 +98,36 @@ const section = navigation.find((s) =>
 );
 const { toc } = useContent();
 const showTocDropdown = ref(false);
+
+let visibleHeadingIds = reactive(new Set());
+let observer = null;
+onMounted(() => {
+  const observerOptions = {
+    threshold: 0,
+    rootMargin: '-72px',
```
Comment (MeganSteenkamp) [inline components/Layout.vue:107]: This means that a section that is below the header would not be considered visible.

File: nuxt.config.js [hunk 2]
```diff
@@ -3,7 +3,16 @@ export default defineNuxtConfig({
     '@nuxt/content'
   ],
   content: {
-    documentDriven: true
+    documentDriven: true,
+    markdown: {
+      remarkPlugins: {
+        'remark-sectionize': {},
```
Comment (MeganSteenkamp) [inline nuxt.config.js:9]: This wraps each heading and the content that follows it in a <section> tag. This is necessary to determine when a section is visible on the page, particularly for a long section where the header is no longer visible.

File: components/Layout.vue [hunk 3]
```diff
@@ -93,6 +102,34 @@ const section = navigation.find((s) =>
 );
 const { toc } = useContent();
 const showTocDropdown = ref(false);
+
+let visibleHeadingId = ref('');
+let observer = null;
+onMounted(() => {
+  const observerOptions = {
+    threshold: 0,
+    rootMargin: '-20% 0px -80% 0px',
```
Comment (MeganSteenkamp) [inline components/Layout.vue:111]: This means that when a section heading is 20% from the top it is considered active.
Note that this means on first page load, when the overview is mainly visible, there is no sections highlighted.
It also means a small section at the bottom may never become highlighted.

File: nuxt.config.js (hunk 2 above)
Comment (AndyClifford) [reply nuxt.config.js:9]: In a section tag?

File: nuxt.config.js [hunk 4]
```diff
@@ -3,7 +3,16 @@ export default defineNuxtConfig({
...
+      },
+      toc: {
```
Comment (AndyClifford) [inline nuxt.config.js:11]: What is this doing?

File: nuxt.config.js (hunk 2 above)
Comment (MeganSteenkamp) [reply nuxt.config.js:9]: Haha yes a section tag - the formatting seems to have actually applied a section tag 😂

File: nuxt.config.js (hunk 4 above)
Comment (MeganSteenkamp) [reply nuxt.config.js:11]: @AndyClifford by default only 2 levels are searched for the TOC.

We previously had a structure like:
- h2
    - h3

We now have a structure like: 
- section
    - h2
        - section
            - h3

Therefore we need to go to 4 levels of depth.

File: components/Layout.vue [hunk 5]
```diff
@@ -22,10 +22,10 @@
       </article>
     </div>
     <nav
-      class="prose-a:no-underline sticky top-[4.5rem] flex flex-col w-full px-4 bg-white drop-shadow-sm lg:max-h-page lg:self-start lg:bg-transparent lg:backdrop-blur-none lg:flex-none lg:overflow-y-auto lg:py-16 lg:col-span-1"
+      class="prose-a:no-underline sticky top-[4.5rem] flex flex-col w-full bg-white drop-shadow-sm lg:max-h-page lg:self-start lg:bg-transparent lg:backdrop-blur-none lg:flex-none lg:overflow-y-auto lg:py-16 lg:col-span-1"
     >
       <button
-        class="px-0 py-3 gap-1 flex items-center focus:outline-none focus:ring-0 mb-0 lg:hidden"
+        class="type-overline text-brand-accent px-6 py-3 gap-1 flex items-center focus:outline-none focus:ring-0 mb-0 lg:hidden"
```
Comment (AndyClifford) [inline components/Layout.vue:28]: Any reason why this has been given `px-6`? It doesn't align with the headings now

image

**Comment by tom-blake** [comment]
@MeganSteenkamp Can you attach mobile and desktop screenshots to this PR?

File: components/Layout.vue (hunk 5 above)
Comment (MeganSteenkamp) [reply components/Layout.vue]: Thanks @AndyClifford that was an incorrect addition, have now removed this.

**Comment by MeganSteenkamp** [comment]
@tom-blake screenshots now added 🚀

**Review by tom-blake — APPROVED** [review]

**Pull request merged**